File: src/main/java/com/hileco/cortex/analysis/edges/EdgeUtility.java
package com.hileco.cortex.analysis.edges;

import com.hileco.cortex.analysis.Graph;
import com.hileco.cortex.analysis.GraphBlock;
import com.hileco.cortex.analysis.GraphNode;
import lombok.Value;

import java.util.List;
import java.util.Optional;
import java.util.stream.Stream;

@Value
public class EdgeUtility<T extends Edge> {
    private Class<T> edgeClass;

    public long count(GraphNode graphNode) {
        return graphNode.getEdges().stream()
                .filter(this.edgeClass::isInstance)
                .count();
    }

    @SuppressWarnings("unchecked")
    public Optional<T> findAny(GraphNode graphNode) {
        return (Optional<T>) graphNode.getEdges().stream()
                .filter(this.edgeClass::isInstance)
                .findAny();
    }

    @SuppressWarnings("unchecked")
    public Optional<T> findAny(Graph graph) {
        return (Optional<T>) graph.getEdges().stream()
                .filter(this.edgeClass::isInstance)
                .findAny();
    }

    @SuppressWarnings("unchecked")
    public Stream<T> filter(Edge edge) {
        return Stream.of(edge)
                .filter(this.edgeClass::isInstance)
                .map(filtered -> (T) filtered);
    }

    public void clear(Graph graph) {
        var edges = graph.getEdges();
        this.clear(edges);
        graph.getGraphBlocks().forEach(this::clear);
    }

    private void clear(GraphBlock graphBlock) {
        var edges = graphBlock.getEdges();
        this.clear(edges);
        graphBlock.getGraphNodes().forEach(this::clear);
    }

    private void clear(GraphNode graphNode) {
        var edges = graphNode.getEdges();
        this.clear(edges);
    }

    private void clear(List<Edge> edges) {
        for (var i = edges.size() - 1; i >= 0; i--) {
            if (this.edgeClass.isInstance(edges.get(i))) {
                edges.remove(i);
            }
        }
    }
}

File: src/main/java/com/hileco/cortex/analysis/edges/EdgeUtility.kt
package com.hileco.cortex.analysis.edges

import com.hileco.cortex.analysis.Graph
import com.hileco.cortex.analysis.GraphBlock
import com.hileco.cortex.analysis.GraphNode
import java.util.stream.Stream

class EdgeUtility<T : Edge>(private val edgeClass: Class<T>) {
    fun count(graphNode: GraphNode): Long {
        return graphNode.edges.stream()
                .filter { edgeClass.isInstance(it) }
                .count()
    }

    fun findAny(graphNode: GraphNode): T? {
        return graphNode.edges.firstOrNull { edgeClass.isInstance(it) } as T?
    }

    fun findAny(graph: Graph): T? {
        return graph.edges.firstOrNull { edgeClass.isInstance(it) } as T?
    }

    fun filter(edge: T): Stream<T> {
        return Stream.of<T>(edge)
                .filter { edgeClass.isInstance(it) }
    }

    fun clear(graph: Graph) {
        val edges = graph.edges
        this.clear(edges)
        graph.graphBlocks.forEach { this.clear(it) }
    }

    private fun clear(graphBlock: GraphBlock) {
        val edges = graphBlock.edges
        this.clear(edges)
        graphBlock.graphNodes.forEach { this.clear(it) }
    }

    private fun clear(graphNode: GraphNode) {
        val edges = graphNode.edges
        this.clear(edges)
    }

    private fun clear(edges: MutableList<Edge>) {
        for (i in edges.indices.reversed()) {
            if (edgeClass.isInstance(edges[i])) {
                edges.removeAt(i)
            }
        }
    }
}
